### PR TITLE
Stream response bodies while sampling audit data

### LINKED ISSUE #4: Proxy buffers response bodies before streaming them to clients
## Summary
The proxy currently reads each upstream response body up to the audit buffering cap before returning the response to the client. For responses under the cap, the full response is buffered before the client receives any bytes. For larger responses, the client still waits until the first capped prefix has been read.

## Impact
This adds avoidable latency and memory pressure on the hot proxy path, especially for large downloads or streaming-like responses. Audit capture should not delay response delivery more than necessary.

## Expected behavior
The proxy should stream response bytes to the client as they arrive while capturing a bounded audit sample in parallel. Audit logging can complete when the response body reaches EOF or is closed.

## Proposed fix
Wrap upstream response bodies with an audit-sampling reader that records a capped prefix while preserving streaming behavior for the client. Log the audit entry when the wrapped body is closed or fully consumed, and add a regression test proving `processRequest` returns before the backend finishes sending a large response.

### PR DESCRIPTION
**Pull request by xbattlax**
## Summary
- stream upstream response bodies directly to clients instead of buffering them in processRequest
- capture a bounded response sample for audit/debug logging as the body is read
- add a regression test proving small responses return before the upstream finishes sending the body

Fixes #4

## Testing
- go test -c -o /tmp/crabtrap-proxy.test ./internal/proxy
- go test ./internal/judge ./internal/llm ./internal/config ./internal/builder ./internal/envutil
- env PATH=/Users/nathanmetzger/go/bin:$PATH make lint
- go test ./internal/proxy -run TestSmallResponseIsStreamedBeforeFullAuditBuffer -count=1 (blocked locally: rootless Docker not found from TestMain/Testcontainers)
- env PATH=/Users/nathanmetzger/go/bin:$PATH make test (lint passed; test phase blocked locally by rootless Docker not found from Testcontainers)

### CODE REVIEW

**Review by bjhaid — CHANGES_REQUESTED** [review]
Thanks for this, sorry can't merge this because of the below

### 1. Audit reliability is fundamentally weakened

The old code wrote the audit entry *before* (or at the same time as) delivering small responses to the client. The new code defers audit logging until the body is fully consumed or closed, meaning:

- If the process crashes (OOM kill, SIGKILL, panic) mid-stream, the audit entry is **permanently lost** with no trace the request was ever approved and forwarded.
- `sync.Once` means if `logEntry()` panics inside `finalize()`, the audit entry is silently dropped and `once.Do` will never retry.
- For long-lived streaming responses (SSE, large downloads), the audit window extends for the entire transfer duration — potentially minutes or hours of unaudited approved traffic.

### 2. Mid-stream read errors no longer produce 502 to the client

The old code (lines 1486-1500 in the base) caught `io.ReadAll` failures and returned a clean `502 Bad Gateway` to the client. The new code has **no error recovery path during streaming**. If the upstream connection drops mid-transfer, the client receives a partial, corrupted response with a 200 status code. The error is only captured in the audit log (if `finalize()` runs). The client has no signal that the response is incomplete.

This is particularly dangerous for API responses where partial JSON could be silently parsed by clients as valid data.

### 3. `*http.Request` reference held for the lifetime of the transfer

`responseAuditBody` captures `request *http.Request` and holds it (along with headers and closures over `h`) until `finalize()` completes. For large/slow downloads, this extends the lifetime of the request object (and its context, body references, TLS state) far beyond what the old code required. This isn't just a memory concern — it means request-scoped resources aren't released promptly.

### 4. Debug logging suppressed for truncated responses

`logResponseDebug` (new code) has this guard:
```go
if len(loggableBody) > 0 && !truncated {
    slog.Debug("response body", ...)
}
```

The old code logged the (truncated) body at debug level even for large responses. The new code suppresses the debug body log entirely when truncated.

### 5. Gzip decompression skipped for truncated responses

`responseBodyForAudit` returns early with the raw (compressed) bytes + truncation marker when `truncated` is true. The old code always attempted gzip decompression on the buffered portion. Audit log consumers expecting readable text in `ResponseBody` for large gzipped responses will now see raw compressed bytes.

### 6. `truncated` flag false-positive on `Close()` without full read

In `finalize()`:
```go
truncated := b.truncated || !b.eof
```

If a caller closes the body without fully reading it (e.g., `serveTLS` write error at line 698-700), `eof` is false, so `truncated` is set to true even if the total response was small. The audit entry will then say "truncated" for a response that wasn't large — it was just interrupted. These are meaningfully different situations for audit analysis but are conflated.

### 7. `captureLocked` uses off-by-one boundary

```go
remaining := maxBufferedBodySize + 1 - len(b.sample)
```

This captures up to `maxBufferedBodySize + 1` bytes into `sample`, matching the old `LimitedReader` behavior for *detection* of oversize. But `responseBodyForAudit` then re-truncates to `maxBufferedBodySize`. The +1 byte is captured, allocated, and immediately discarded.

### 8. No pre-allocation of sample buffer

`captureLocked` grows `b.sample` via repeated `append` for up to 10MB. The old code used a single `io.ReadAll` with `LimitedReader`, which Go's runtime can optimize. For responses near the 10MB cap, the new code will trigger multiple slice reallocations.

### 9. No test that audit entries are actually written

`TestSmallResponseIsStreamedBeforeFullAuditBuffer` verifies the response arrives before the backend finishes sending. It does **not** verify the audit entry was logged. The whole point of this refactor is to preserve audit reliability while streaming — that property is untested.

A test should:
- Stream a response
- Fully read and close the body
- Assert the audit JSONL file contains the expected entry

### 10. No test for the crash/partial-read audit path

What happens if the body is closed without being fully read? What does the audit entry look like? What if `Close()` is called concurrently with `Read()`? These edge cases are where the new code diverges most from the old behavior and are the most likely source of bugs.

**Comment by xbattlax** [comment]
Thanks, I pushed a follow-up that narrows the streaming path and restores early audit persistence for the cases we still stream.

What changed:
- kept the original fully-buffered behavior for normal responses under the cap
- limited immediate streaming to large responses with `Content-Length` above the cap and to `text/event-stream` responses
- wrote the audit entry before returning streamed responses, then added a DB-side response update when the streamed body completes or closes
- removed the request retention from the streaming wrapper and restored truncated debug logging / gzip prefix decoding for streamed samples
- updated the regression tests to cover large known-length responses, event streams, and early audit entry creation

I also re-ran lint locally and re-ran `make test`; lint passed, while the full suite still hits the same local Testcontainers/Docker limitation as before.

File: internal/proxy/handler.go [hunk 1]
```diff
@@ -1632,6 +1853,28 @@ func (h *Handler) logEntry(entry types.AuditEntry) {
...
+		slog.Error("failed to update streamed response audit", "request_id", requestID, "error", err)
+	}
+}
+
+func shouldStreamResponse(resp *http.Response) bool {
+	if resp.ContentLength > maxBufferedBodySize {
+		return true
+	}
+	contentType := strings.ToLower(resp.Header.Get("Content-Type"))
+	return strings.Contains(contentType, "text/event-stream")
```
Comment (bjhaid) [inline internal/proxy/handler.go:1882]: this should also do check if resp.ContentLength == -1 or resp.TransferEncoding contains "chunked"

**Review by bjhaid — COMMENTED** [review]
one small feedback and we should be ready to merge. THanks!

**Comment by xbattlax** [comment]
Thanks, addressed in aadb7b0.

The streaming gate now also treats unknown-length responses (ContentLength == -1) and explicit Transfer-Encoding: chunked responses as streaming candidates, alongside large known-length responses and text/event-stream. I also added a chunked-response regression test that verifies processRequest returns before the upstream finishes the body.

Local verification:
- go test -c -o /tmp/crabtrap-proxy.test ./internal/proxy passed
- go test ./internal/proxy -run TestChunkedResponseIsStreamedBeforeCompletion -count=1 is still blocked locally by Testcontainers/rootless Docker before the selected test can run

**Comment by xbattlax** [comment]
Addressed the remaining Greptile note in 615fdde.

Ledger update:
- Greptile: UpdateResponse now returns a contextual marshal response headers error instead of discarding json.Marshal(responseHeaders). The existing caller logs UpdateResponse failures through slog.Error, so unexpected marshal failures are observable and no silent null header payload is written.
- bjhaid inline feedback on unknown-length and chunked streaming remains addressed by aadb7b0.

Local verification:
- go test -c -o /tmp/crabtrap-admin.test ./internal/admin
- go test -c -o /tmp/crabtrap-proxy.test ./internal/proxy
- go vet ./...
- /Users/nathanmetzger/go/bin/staticcheck ./...
- git diff --check

Full runtime test execution is still limited locally by the existing Testcontainers/rootless Docker issue, so CI remains the source of truth for the full go test -race -p 1 ./... pass.

**Comment by xbattlax** [comment]
Addressed the follow-up Greptile gzip audit note in 0c3459d.

Ledger update:
- Greptile: responseBodyForAudit now propagates the decoded gzip truncation flag returned by decompressGzipResponsePrefix, so streamed gzip samples that exceed the decompressed audit cap get the decompressed response body truncated marker instead of silently omitting it.
- Added TestResponseBodyForAuditMarksGzipDecompressionTruncation to cover this behavior in CI.

Local verification:
- go test -c -o /tmp/crabtrap-admin.test ./internal/admin
- go test -c -o /tmp/crabtrap-proxy.test ./internal/proxy
- go vet ./...
- /Users/nathanmetzger/go/bin/staticcheck ./...
- git diff --check

Local execution of the new proxy test is still blocked before test selection by internal/proxy.TestMain requiring rootless Docker/Testcontainers, so CI remains the source of truth for that runtime test.

**Comment by xbattlax** [comment]
Addressed the latest Greptile note in 656066e.

Ledger update:
- Greptile: the broad ContentLength == -1 streaming trigger is intentional. Unknown-length responses cannot be safely bounded against maxBufferedBodySize without reading to EOF first, which is exactly what caused the buffering issue this PR fixes. This also matches bjhaid’s inline request to stream ContentLength == -1 or Transfer-Encoding: chunked responses.
- I added a code comment next to the ContentLength == -1 branch to make that intent explicit for future reviewers.

Local verification:
- go test -c -o /tmp/crabtrap-proxy.test ./internal/proxy
- go vet ./...
- /Users/nathanmetzger/go/bin/staticcheck ./...
- git diff --check

**Comment by xbattlax** [comment]
Addressed the latest Greptile feedback in de39fce.

Ledger update:
- Greptile issue 1: UpdateResponse now uses a 5 second context timeout, so a hung Postgres call cannot block the streamed response finalization path indefinitely.
- Greptile issue 2: UpdateResponse is now part of admin.AuditReaderIface. That removes the silent non-updater auditReader case for any non-nil audit reader passed into the proxy. The proxy now calls h.auditReader.UpdateResponse directly, and only logs a warning when no audit reader is configured at all.
- The ContentLength == -1 streaming behavior remains intentional and matches bjhaid’s inline request; this change makes the two-phase audit update contract explicit instead of optional.

Local verification:
- go test -c -o /tmp/crabtrap-admin.test ./internal/admin
- go test -c -o /tmp/crabtrap-proxy.test ./internal/proxy
- go vet ./...
- /Users/nathanmetzger/go/bin/staticcheck ./...
- git diff --check

**Review by bjhaid — APPROVED** [review]
Thanks!
